common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java

```diff
@@ -0,0 +1,249 @@
+     *
+     * @param fileToWatch path of the file to watch
+     * @param callback          the callback to invoke when a modification to the {@code fileToWatch} is detected
+     * @param loopContinuously  whether to keep continue to look for file modification after one iteration. This option
+     *                          is useful for testing only.
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     * @throws NullPointerException if either {@code fileToWatch} or {@code callback} is null
+     */
+    @VisibleForTesting
+    FileModificationEventWatcher(@NonNull Path fileToWatch, @NonNull Consumer<WatchEvent<?>> callback,
+                                 boolean loopContinuously, boolean checkForFileExistence)
+            throws FileNotFoundException {
+
+        // Set the name for this object/thread for identification purposes.
+        super("pravega-file-watcher-" + THREAD_NUM.incrementAndGet());
+        Exceptions.checkNotNullOrEmpty(fileToWatch.toString(), "fileToWatch");
+        if (checkForFileExistence && !fileToWatch.toFile().exists()) {
+            throw new FileNotFoundException(String.format("File [%s] does not exist.", fileToWatch));
+        }
+
+        this.watchedFilePath = fileToWatch;
+        this.callback = callback;
+        this.loopContinuously = loopContinuously;
+        setUncaughtExceptionHandler(uncaughtExceptionalHandler);
+    }
+
+    @VisibleForTesting
+    String getWatchedFileName() {
+        Path fileName = this.watchedFilePath.getFileName();
+        if (fileName != null) {
+            return fileName.toString();
+        } else {
+            // It should never happen that file name is null, but we have to perform this defensive null check to
+            // satisfy SpotBugs. File name will be null for a path like "/".
+            throw new IllegalStateException("File name is null");
+        }
+    }
+
+    @VisibleForTesting
+    Path getWatchedDirectory() {
+        return this.watchedFilePath.getParent();
+    }
+
+    @Override
+    @SuppressWarnings("SleepWhileInLoop")
+    public void run() {
+        WatchKey watchKey = null;
+        WatchService watchService = null;
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            log.debug("Done creating watch service for watching file at path: {}", this.watchedFilePath);
+
+            String fileName = getWatchedFileName();
+            Path directoryPath = getWatchedDirectory();
+            log.debug("Directory being watched is {}", directoryPath);
+
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE);
+            log.debug("Done setting up watch for modify entries for file at path: {}", this.watchedFilePath);
```

Comment:
The message in the debug log message does not read well, please rephrase as I'm not what it is trying to say other than that it has set up the watch.

Reply:
I've fixed it now.